### DISCUSSION TITLE
when collectionID start with 0x0 the payload remove the 0 after x

### DISCUSSION THREAD

**Discussion by spielcrypto** [aptos-labs/aptos-developer-discussions#422] [Questions] [answered]
### Discord user ID

spielcrypto

### Describe your question in detail.

I try to set a collection Id calling the function set_config of this [module](https://explorer.aptoslabs.com/account/0x9ed8aa766c59553a345523a1840e4727af9e16d6b9b98c0defdea2f639bc0e63/modules/run/burn/set_config?network=testnet)
the id is `0x0ed55bdab2e8d49cf51df95c4c10382be09353ea68260c89ac342b74dbaa8104` and when I call to the wallet it shows a correct collectionId as is possible to appreciate in the image bellow

![image](https://github.com/user-attachments/assets/deeb268e-71a5-40b6-b9ec-76a34f263f78)

however in the payload the id change removing the 0 after x as is possible to see [here](https://explorer.aptoslabs.com/txn/0xf513ab3b2b21951a0bc4f01210eae941b0da7c52c46ae2c01a3dfd7a97e4bee5/payload?network=testnet) sending collection id param to the contract like this `0xed55bdab2e8d49cf51df95c4c10382be09353ea68260c89ac342b74dbaa8104`

### What error, if any, are you getting?
* When collectionID start with 0x0 the payload remove the 0 after x
* The issue happens in testnet and mainnet
* Maybe the issue happens in wallet adapter

### What have you tried or looked at?  Or how can we reproduce the error?
it is explained above in the description

### Which operating system are you using?

Linux (Ubuntu, Fedora, Windows WSL, etc.)

### Which SDK or tool are you using? (if any)

TypeScript SDK

### Describe your environment or tooling in detail

Aptos client 4.1.0
Testnet and Mainnet

**Comment by banool** [comment]
Sadly for backwards compatibility reasons a lot of our APIs, tools, etc. do not conform to [AIP-40](https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-40.md). This means that sometimes certain services, such as the node API, use this incorrect representation where they trim the padding zeroes (AIP-40 says this is only acceptable for special addresses like 0x1).

In your case, the data is actually correct on chain, it is just that the node API is converting the data into a string incorrectly (removing the padding zeroes). The explorer (https://explorer.aptoslabs.com/txn/0xf513ab3b2b21951a0bc4f01210eae941b0da7c52c46ae2c01a3dfd7a97e4bee5/payload?network=testnet) is just displaying what the node API returns verbatim (https://fullnode.testnet.aptoslabs.com/v1/transactions/by_hash/0xf513ab3b2b21951a0bc4f01210eae941b0da7c52c46ae2c01a3dfd7a97e4bee5).

If the on chain type is `address`, you can feel confident that even if the representation from the API looks wrong, the data is correct on chain. 

We have some ideas for how to fix this, but it's tricky because we also need to not break the API for existing users, even if the current behavior is wrong.

@gregnazario might want to chime in with more details.

**Reply by spielcrypto** [reply]
Thanks for the reply!

The problem of this issue is that after in frontend I want to compare if user hold that collection but doesn't match because for the issue above, for now I will remove also the 0 after x, I don't like such solution but I hope it is temporal

**Reply by banool** [reply, marked as answer]
Yeah hopefully we can fix this, but in the near term, I recommend using SDK code to help with this. For example, in Typescript:
```ts
import { AccountAddress } from "@aptos-labs/ts-sdk";

const addressOne = AccountAddress.fromString("0x123");
const addressTwo = AccountAddress.fromString("0x123");
addressOne.equals(addressTwo);
```
